src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -310,33 +331,59 @@ internal override sealed bool TryReadAsObject(ref Utf8JsonReader reader, JsonSer
         }
 
         /// <summary>
-        /// Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Performance optimization. Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Although the JIT added support to detect this IL pattern (box+isinst+br), it still manifests in TryWrite().
+        /// </summary>
+        internal virtual bool IsNull(in T value) => value is null;
```

Comment:
Ran a few experiments locally and it seems like the `in` modifier is the culprit for the boxing behavior. Changing to
```
private bool IsNull(T value) => value is null;
```
makes the regression go away. It comes at the cost of having to pass the parameter by value but it's negligible compared to boxing. This also avoids needing to make a virtual call. cc @EgorBo

Reply:
Thanks. Based on this and your other comment around using `default(T) is null` in that LOC in question, I re-ran some micro benchmarks and found, as expected, that value types such as `int` are faster, while reference types including `string` will be a bit slower. However, removing the extra caching and make the code simpler is also a win, so I plan on making these changes in the next commit and re-running the full benchmarks again.

Reply:
Thanks. Based on this and your other comment around using default(T) is null in that LOC in question, I re-ran some micro benchmarks and found, as expected, that value types such as int are faster, while reference types including string will be a bit slower. However, removing the extra caching and make the code simpler is also a win, so I plan on making these changes in the next commit and re-running the full benchmarks again.

Reply:
@eiriktsarpalis how can I reproduce the boxing? e.g. I tried [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGufgIEADYMAkrgByAVyFCAPABUAfAAoAlgDsGChgDdsQqTACUDALxK9BowzW4GGmUIDcfbm4YDh2mLgzkVAHEpNQATAH4HUwsxSScVDWNXah5PQREFXww6IJDQqPNLcWlZBKSPLwys0nUNDEjEwtiSoTLk1JoAXyA=) and don't see any boxing.

Reply:
Could [this](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuHgIIANgwCSuAHIBXAQIA8AFQB8ACgCWAOwZyGAN2wCJMAJQMAvAp16DDFbgZqpAgNxceL7gG0AygAtsUAA4AMtjAAHQAShJqGCr4MKEAUioYAOIwajBQKmBKGACe/jAQAGaq0YaGALpuDF6+AcFhkdGx8Ump6ZnZuQVFpeoYAPwV1dQ8vPxCnhBxAPIY3pkAsjALEAAm8srqmhb6RqbmopLSSrr7hs7UAL5AA=) be a reproduction? Not sure what `L000e: call 0x00007ffd016a1b10` is doing but seems consistent with what I'm seeing when I profile the repro:

![image](https://user-images.githubusercontent.com/2813363/129445196-1d39c330-b899-4761-853c-af3de640d7c3.png)

Reply:
FWIW here's a commit that reproduces the boxing behavior: https://github.com/eiriktsarpalis/runtime/commit/a44c0eecf099b8aa6a119fc2d4abd9a904b4f700

It can be observed running the benchmark added here https://github.com/dotnet/performance/pull/1920